File: C++/Prototype/ShapeFactory/shape.cc
#include <iostream>
#include "shape.h"


Shape::Shape(const Shape &) {
    // Some copy operations
}


Rectangle::Rectangle(const Rectangle &) {
    // Some copy operations
}

void Rectangle::draw() {
    std::cout << "Drawing a rectangle...\n";
}

std::unique_ptr<Shape> Rectangle::clone() {
    return std::make_unique<Rectangle>(*this);
}


Triangle::Triangle(const Triangle &) {
    // Some copy operations
}

void Triangle::draw() {
    std::cout << "Drawing a triangle...\n";
}

std::unique_ptr<Shape> Triangle::clone() {
    return std::make_unique<Triangle>(*this);
}

File: Java/Prototype/ShapeFactory/src/com/ericgtkb/Shape.java
package com.ericgtkb;


// This is the prototype interface, all subclass must override clone
public abstract class Shape implements Cloneable {
    public abstract void draw();

    // For complex objects deep copy may be required
    @Override
    public Shape clone() throws CloneNotSupportedException {
        return (Shape) super.clone();
    }
}

class Rectangle extends Shape {
    @Override
    public void draw() {
        System.out.println("Drawing a rectangle...");
    }

    @Override
    public Shape clone() throws CloneNotSupportedException {
        return (Rectangle) super.clone();
    }
}

class Triangle extends Shape {
    @Override
    public void draw() {
        System.out.println("Drawing a triangle...");
    }

    @Override
    public Shape clone() throws CloneNotSupportedException {
        return (Triangle) super.clone();
    }
}
